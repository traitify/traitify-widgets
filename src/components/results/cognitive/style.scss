@import "style/helpers";

.container {
  @extend %container;

  text-align: center;

  a, a:visited {
    text-decoration: none;

    @include theme("color", "theme-color");
  }
  img {
    margin: 0 auto;
    max-width: 200px;
    width: 50%;
  }
}
.heading {
  @extend %heading;

  @include theme("color", "theme-color");
}
