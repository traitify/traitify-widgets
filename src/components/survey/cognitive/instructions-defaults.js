const defaultAnswers = [4, 1, 2];
const urlBase = "https://cdn.traitify.com/images/cognitive";

export const defaultExamples = [
  ({timeTrial, translate}) => ({
    button: translate("cognitive_instructions_step_1_button"),
    heading: translate("cognitive_instructions_step_1_heading"),
    text: translate(timeTrial ? "cognitive_instructions_trial_step_1_text" : "cognitive_instructions_step_1_text")
  }),
  ({translate, type}) => ({
    button: translate("cognitive_instructions_step_2_button"),
    heading: translate("cognitive_instructions_step_2_heading"),
    text: translate("cognitive_instructions_step_2_text"),
    video: `${urlBase}/practice-example-${type}.mp4`
  })
];

export const defaultExplanations = defaultAnswers.map((_, index) => (
  ({translate, type}) => ({
    button: translate(`cognitive_practice_step_${index + 1}_button`),
    heading: translate(`cognitive_practice_step_${index + 1}_heading`),
    text: translate(`cognitive_practice_step_${index + 1}_text`),
    video: `${urlBase}/practice-${index + 1}-${type}.mp4`
  })
));

export const defaultInstruction = ({minimal, timeTrial, timed, translate}) => ({
  button: translate("cognitive_instructions_step_4_button"),
  heading: translate("cognitive_instructions_step_4_heading"),
  text: translate(
    timeTrial
      ? `cognitive_instructions_trial_step_4_${timed ? "timed" : "untimed"}${minimal ? "_minimal" : ""}_html`
      : "cognitive_instructions_step_4_html"
  )
});

export const defaultQuestions = defaultAnswers.map((answer, index) => ({
  correctAnswerID: `r-${index}-${answer}`,
  id: `s-${index}`,
  questionImage: {url: `https://cdn.traitify.com/images/cognitive/practice-question-${index + 1}/question.png`},
  responses: [1, 2, 3, 4].map((response) => ({
    id: `r-${index}-${response}`,
    image: {url: `https://cdn.traitify.com/images/cognitive/practice-question-${index + 1}/response-${response}.png`}
  }))
}));
