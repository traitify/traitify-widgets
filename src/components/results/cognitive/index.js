import DangerousHTML from "components/common/dangerous-html";
import useTranslate from "lib/hooks/use-translate";
import style from "./style.scss";

export default function Cognitive() {
  const translate = useTranslate();

  return (
    <section className={style.container}>
      <img alt="Brain" src="https://cdn.traitify.com/images/cognitive/brain.png" />
      <div className={style.heading}>{translate("cognitive_results_heading")}</div>
      <DangerousHTML html={translate("cognitive_results_html")} />
    </section>
  );
}
