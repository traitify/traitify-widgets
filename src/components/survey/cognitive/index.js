/* eslint-disable no-alert */
import {useEffect, useRef, useState} from "react";
import {useRecoilRefresher_UNSTABLE as useRecoilRefresher} from "recoil";
import Loading from "components/common/loading";
import useAssessment from "lib/hooks/use-assessment";
import useCache from "lib/hooks/use-cache";
import useCacheKey from "lib/hooks/use-cache-key";
import useComponentEvents from "lib/hooks/use-component-events";
import useDidUpdate from "lib/hooks/use-did-update";
import useGraphql from "lib/hooks/use-graphql";
import useHttp from "lib/hooks/use-http";
import useOption from "lib/hooks/use-option";
import useTranslate from "lib/hooks/use-translate";
import {cognitiveAssessmentQuery} from "lib/recoil";
import Instructions from "./instructions";
import Slide from "./slide";
import Timer from "./timer";
import {useQuestionsLoader} from "./helpers";
import style from "./style.scss";

export default function Cognitive() {
  const assessment = useAssessment("cognitive");
  const assessmentCacheKey = useCacheKey("assessment");
  const cache = useCache();
  const cacheKey = useCacheKey({scope: ["slides"], type: "assessment"});
  const graphQL = useGraphql();
  const http = useHttp();
  const options = useOption("survey") || {};
  const refreshAssessment = useRecoilRefresher(cognitiveAssessmentQuery);
  const translate = useTranslate();

  const [initialQuestions, setInitialQuestions] = useState([]);
  const {dispatch, questions} = useQuestionsLoader(initialQuestions);
  const [disability, setDisability] = useState(false);
  const [onlySkipped, setOnlySkipped] = useState(false);
  const [questionIndex, setQuestionIndex] = useState(null);
  const [skipped, setSkipped] = useState(null);
  const [startTime, setStartTime] = useState(null);
  const [submitAttempts, setSubmitAttempts] = useState(0);
  const submitting = useRef(false);
  const state = {
    disability,
    dispatch,
    initialQuestions,
    questions,
    onlySkipped,
    questionIndex,
    skipped,
    startTime
  };
  const onSelect = (answer) => {
    dispatch({answer, questionIndex, type: "response"});
    if(!onlySkipped) { return setQuestionIndex(questionIndex + 1); }

    const newQuestionIndex = questions
      .findIndex((question, index) => index > questionIndex && !question.answer.answerId);
    setQuestionIndex(newQuestionIndex === -1 ? questions.length + 1 : newQuestionIndex);
  };
  const onStart = (_options) => {
    if(_options.disability) { setDisability(true); }

    setQuestionIndex(0);
    setStartTime(Date.now());
  };
  const onSubmit = () => {
    if(assessment.completed) { return; }
    if(submitAttempts > 3) { return; }
    if(submitting.current) { return; }

    submitting.current = true;

    const query = graphQL.cognitive.update;
    const variables = {
      answers: questions.map((question) => (
        question.answer ? ({
          answerId: question.answer.answerId,
          questionId: question.id,
          skipped: question.answer.skipped || false,
          timeTaken: question.answer.timeTaken < 1 ? 1 : question.answer.timeTaken
        }) : ({
          questionId: question.id,
          skipped: true, // Currently required by API
          timeTaken: 1 // Currently required by API
        })
      )),
      learningDisability: disability,
      overallTimeTaken: startTime ? parseInt((Date.now() - startTime) / 1000, 10) : 1000,
      testID: assessment.id
    };

    http.post(graphQL.cognitive.path, {query, variables}).then(({data, errors}) => {
      if(!errors && data.completeCognitiveTest.success) {
        cache.set(assessmentCacheKey, {...assessment, completed: true});
        refreshAssessment();

        submitting.current = false;
      } else {
        console.warn(errors || data); // eslint-disable-line no-console

        submitting.current = false;

        setTimeout(() => setSubmitAttempts((x) => x + 1), 2000);
      }
    });
  };

  useComponentEvents("Survey", {...state});
  useDidUpdate(() => { onSubmit(); }, [submitAttempts]);
  useEffect(() => {
    if(!assessment) { return; }

    const cachedData = cache.get(cacheKey);
    if(cachedData) {
      setDisability(cachedData.disability);
      setOnlySkipped(cachedData.onlySkipped);
      setSkipped(cachedData.skipped);

      setQuestionIndex(cachedData.questionIndex);
      setInitialQuestions(cachedData.questions);
      setStartTime(cachedData.startTime);
    } else {
      setInitialQuestions(assessment.questions);
    }
  }, [
    assessment && assessment.questions,
    assessment && assessment.localeKey
  ]);

  useEffect(() => {
    if(!assessment) { return; }
    if(questions.length === 0) { return; }

    cache.set(cacheKey, {
      disability,
      onlySkipped,
      questions,
      questionIndex,
      skipped,
      startTime
    });
  }, [disability, onlySkipped, questions, questionIndex, skipped, startTime]);

  useEffect(() => {
    if(questions.length === 0) { return; }
    if(questions.length > questionIndex) { return; }
    if(onlySkipped) { return onSubmit(); }

    setOnlySkipped(true);

    const skippedIndexes = questions
      .map(({answer}, index) => ({answer, index}))
      .filter(({answer}) => !answer.answerId)
      .map(({index}) => index);

    if(skippedIndexes.length === 0) { return onSubmit(); }
    if(window.confirm(translate("cognitive_confirm_retry"))) {
      setSkipped(skippedIndexes);
      setQuestionIndex(skippedIndexes[0]);
    } else {
      return onSubmit();
    }
  }, [questions, questionIndex]);
  const question = questions[questionIndex];
  const nextQuestion = questions[questionIndex + 1];

  if(!assessment) { return null; }
  if(assessment.completed) { return null; }
  if(questionIndex === null) {
    return (
      <Instructions
        captureLearningDisability={options.captureLearningDisability}
        initialLearningDisability={options.initialLearningDisability}
        onStart={onStart}
        options={options}
        translate={translate}
      />
    );
  }
  if(!question || submitting.current) { return <Loading />; }

  const index = skipped ? skipped.indexOf(questionIndex) : questionIndex;
  const total = skipped ? skipped.length : questions.length;
  const progress = (100.0 * (index + 1)) / total;

  return (
    <div className={style.container}>
      <div className={style.statusContainer}>
        {!options.disableTimeLimit && (
          <Timer
            onFinish={onSubmit}
            startTime={startTime}
            timeAllowed={
              disability
                ? options.specialTimeLimit || assessment.specialAllottedTime
                : options.timeLimit || assessment.allottedTime
            }
          />
        )}
        <div className={style.status}>
          {skipped && <span>{translate("cognitive_skipped_questions")} </span>}
          <span>{index + 1} / {total}</span>
        </div>
        <div className={style.progressBar}>
          <div className={style.progress} style={{width: `${progress}%`}} />
        </div>
      </div>
      <Slide
        onSelect={onSelect}
        question={question}
        translate={translate}
      />
      {nextQuestion && (
        <Slide
          className={style.hide}
          onSelect={onSelect}
          question={nextQuestion}
          translate={translate}
        />
      )}
    </div>
  );
}
