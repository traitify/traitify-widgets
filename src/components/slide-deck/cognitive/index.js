/* eslint-disable no-alert */
import PropTypes from "prop-types";
import {useEffect, useRef, useState} from "react";
import Loading from "components/loading";
import {update as updateQuery} from "lib/graphql/queries/cognitive";
import TraitifyPropTypes from "lib/helpers/prop-types";
import useDidMount from "lib/hooks/use-did-mount";
import useDidUpdate from "lib/hooks/use-did-update";
import withTraitify from "lib/with-traitify";
import Instructions from "./instructions";
import Slide from "./slide";
import Timer from "./timer";
import {useQuestionsLoader} from "./helpers";
import style from "./style.scss";

function Cognitive(props) {
  const {
    assessment,
    cache,
    getCognitiveAssessment,
    getOption,
    isReady,
    traitify,
    translate,
    ui
  } = props;
  const disableTimeLimit = getOption("slideDeck", "disableTimeLimit");
  const [initialQuestions, setInitialQuestions] = useState([]);
  const {dispatch, questions} = useQuestionsLoader(initialQuestions);
  const [disability, setDisability] = useState(false);
  const [onlySkipped, setOnlySkipped] = useState(false);
  const [questionIndex, setQuestionIndex] = useState(null);
  const [skipped, setSkipped] = useState(null);
  const [startTime, setStartTime] = useState(null);
  const [submitAttempts, setSubmitAttempts] = useState(0);
  const submitting = useRef(false);
  const state = {
    disability,
    dispatch,
    initialQuestions,
    questions,
    onlySkipped,
    questionIndex,
    skipped,
    startTime
  };
  const onSelect = (answer) => {
    dispatch({answer, questionIndex, type: "response"});
    if(!onlySkipped) { return setQuestionIndex(questionIndex + 1); }

    const newQuestionIndex = questions
      .findIndex((question, index) => index > questionIndex && !question.answer.answerId);
    setQuestionIndex(newQuestionIndex === -1 ? questions.length + 1 : newQuestionIndex);
  };
  const onStart = (options) => {
    if(options.disability) { setDisability(true); }

    setQuestionIndex(0);
    setStartTime(Date.now());
  };
  const onSubmit = () => {
    if(isReady("results")) { return; }
    if(submitAttempts > 3) { return; }
    if(submitting.current) { return; }

    submitting.current = true;

    const answers = questions.map((question) => (
      question.answer ? ({
        answerId: question.answer.answerId,
        questionId: question.id,
        skipped: question.answer.skipped || false,
        timeTaken: question.answer.timeTaken < 1 ? 1 : question.answer.timeTaken
      }) : ({
        questionId: question.id,
        skipped: true, // Currently required by API
        timeTaken: 1 // Currently required by API
      })
    ));
    const query = updateQuery({
      params: {
        answers,
        learningDisability: disability,
        overallTimeTaken: parseInt((Date.now() - startTime) / 1000, 10),
        testId: assessment.id
      }
    });

    traitify.post("/cognitive-tests/graphql", query).then(({data, errors}) => {
      if(!errors && data.completeCognitiveTest.success) {
        ui.trigger("SlideDeck.finished", {props, state});
        // If actual results are shown, use this, but for now it won't work because the API is async
        // getCognitiveAssessment({force: true});
        const key = `${assessment.localeKey.toLowerCase()}.cognitive-assessment.${assessment.id}`;
        cache.set(key, {...assessment, completed: true});
        getCognitiveAssessment();
      } else {
        console.warn(errors || data); // eslint-disable-line no-console
        submitting.current = false;
        setSubmitAttempts(submitAttempts + 1);
      }
    });
  };

  useDidMount(() => { ui.trigger("SlideDeck.initialized", {props, state}); });
  useDidUpdate(() => { ui.trigger("SlideDeck.updated", {props, state}); });
  useDidUpdate(() => { onSubmit(); }, [submitAttempts]);
  useEffect(() => {
    if(!assessment) { return; }

    const cachedData = cache.get(`cognitive.slide-deck.${assessment.id}`);
    if(cachedData) {
      setDisability(cachedData.disability);
      setOnlySkipped(cachedData.onlySkipped);
      setSkipped(cachedData.skipped);

      setQuestionIndex(cachedData.questionIndex);
      setInitialQuestions(cachedData.questions);
      setStartTime(cachedData.startTime);
    } else {
      setInitialQuestions(assessment.questions);
    }
  }, [
    assessment && assessment.questions,
    assessment && assessment.localeKey
  ]);

  useEffect(() => {
    if(!assessment) { return; }
    if(questions.length === 0) { return; }

    cache.set(`cognitive.slide-deck.${assessment.id}`, {
      disability,
      onlySkipped,
      questions,
      questionIndex,
      skipped,
      startTime
    });
  }, [disability, onlySkipped, questions, questionIndex, skipped, startTime]);

  useEffect(() => {
    if(questions.length === 0) { return; }
    if(questions.length > questionIndex) { return; }
    if(onlySkipped) { return onSubmit(); }

    setOnlySkipped(true);

    const skippedIndexes = questions
      .map(({answer}, index) => ({answer, index}))
      .filter(({answer}) => !answer.answerId)
      .map(({index}) => index);

    if(skippedIndexes.length === 0) { return onSubmit(); }
    if(window.confirm(translate("cognitive_confirm_retry"))) {
      setSkipped(skippedIndexes);
      setQuestionIndex(skippedIndexes[0]);
    } else {
      return onSubmit();
    }
  }, [questions, questionIndex]);
  const question = questions[questionIndex];
  const nextQuestion = questions[questionIndex + 1];

  if(isReady("results")) { return null; }
  if(questionIndex === null) {
    return (
      <Instructions
        captureLearningDisability={getOption("slideDeck", "captureLearningDisability")}
        initialLearningDisability={getOption("slideDeck", "initialLearningDisability")}
        onStart={onStart}
        options={getOption("slideDeck") || {}}
        translate={translate}
      />
    );
  }
  if(!question || submitting.current) { return <Loading />; }

  const index = skipped ? skipped.indexOf(questionIndex) : questionIndex;
  const total = skipped ? skipped.length : questions.length;
  const progress = (100.0 * (index + 1)) / total;

  return (
    <div className={style.container}>
      <div className={style.statusContainer}>
        {!disableTimeLimit && (
          <Timer
            onFinish={onSubmit}
            startTime={startTime}
            timeAllowed={
              disability
                ? getOption("slideDeck", "specialTimeLimit") || assessment.specialAllottedTime
                : getOption("slideDeck", "timeLimit") || assessment.allottedTime
            }
          />
        )}
        <div className={style.status}>
          {skipped && <span>{translate("cognitive_skipped_questions")} </span>}
          <span>{index + 1} / {total}</span>
        </div>
        <div className={style.progressBar}>
          <div className={style.progress} style={{width: `${progress}%`}} />
        </div>
      </div>
      <Slide
        className={style.slideContainer}
        onSelect={onSelect}
        question={question}
        translate={translate}
      />
      {nextQuestion && (
        <Slide
          className={style.hide}
          onSelect={onSelect}
          question={nextQuestion}
          translate={translate}
        />
      )}
    </div>
  );
}

Cognitive.defaultProps = {assessment: null};
Cognitive.propTypes = {
  assessment: PropTypes.shape({
    allottedTime: PropTypes.number.isRequired,
    id: PropTypes.string.isRequired,
    localeKey: PropTypes.string.isRequired,
    questions: PropTypes.arrayOf(
      PropTypes.shape({id: PropTypes.string.isRequired})
    ).isRequired,
    specialAllottedTime: PropTypes.number.isRequired
  }),
  cache: PropTypes.shape({
    get: PropTypes.func.isRequired,
    set: PropTypes.func.isRequired
  }).isRequired,
  getCognitiveAssessment: PropTypes.func.isRequired,
  getOption: PropTypes.func.isRequired,
  isReady: PropTypes.func.isRequired,
  traitify: TraitifyPropTypes.traitify.isRequired,
  translate: PropTypes.func.isRequired,
  ui: TraitifyPropTypes.ui.isRequired
};

export {Cognitive as Component};
export default withTraitify(Cognitive);
