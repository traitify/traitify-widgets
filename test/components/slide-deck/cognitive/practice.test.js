import Component from "components/slide-deck/cognitive/practice";
import Slide from "components/slide-deck/cognitive/slide";
import {defaultExplanations, defaultQuestions} from "components/slide-deck/cognitive/instructions-defaults";
import ComponentHandler, {act} from "support/component-handler";
import useResizeMock from "support/hooks/use-resize-mock";
import useWindowMock from "support/hooks/use-window-mock";

jest.mock("components/slide-deck/cognitive/slide", () => (() => <div className="mock">Slide</div>));

const selectAnswer = ({component, index}) => {
  const {onSelect, question} = component.instance.findByType(Slide).props;

  onSelect({answerId: question.responses[index].id, timeTaken: 200});
};

describe("Practice", () => {
  let props;

  beforeEach(() => {
    props = {
      onFinish: jest.fn().mockName("onFinish"),
      practiceExplanations: defaultExplanations,
      practiceQuestions: defaultQuestions,
      translate: jest.fn().mockName("translate").mockImplementation((value) => value)
    };
  });

  describe("actions", () => {
    useWindowMock("alert");

    it("finishes", () => {
      const component = new ComponentHandler(<Component {...props} />);
      component.act(() => selectAnswer({component, index: 3}));
      component.act(() => component.findByText("cognitive_practice_step_1_button").props.onClick());
      component.act(() => selectAnswer({component, index: 3}));
      component.act(() => component.findByText("cognitive_practice_step_2_button").props.onClick());
      component.act(() => selectAnswer({component, index: 3}));
      component.act(() => component.findByText("cognitive_practice_step_3_button").props.onClick());

      expect(component.tree).toMatchSnapshot();
      expect(props.onFinish).toHaveBeenCalled();
    });

    it("doesn't skip", () => {
      const component = new ComponentHandler(<Component {...props} />);
      component.instance.findByType(Slide).props.onSkip();

      expect(window.alert).toHaveBeenCalled();
    });
  });

  describe("videos", () => {
    let component;

    useResizeMock();

    describe("horizontal", () => {
      beforeEach(() => {
        act(() => window.resizeTo(1200, 800));
        component = new ComponentHandler(<Component {...props} />);
        component.act(() => selectAnswer({component, index: 0}));
      });

      it("renders component", () => {
        expect(component.tree).toMatchSnapshot();
      });

      it("updates to vertical", () => {
        component.act(() => window.resizeTo(600, 800));

        expect(component.tree).toMatchSnapshot();
      });
    });

    describe("vertical", () => {
      beforeEach(() => {
        act(() => window.resizeTo(600, 800));
        component = new ComponentHandler(<Component {...props} />);
        component.act(() => selectAnswer({component, index: 0}));
      });

      it("renders component", () => {
        expect(component.tree).toMatchSnapshot();
      });

      it("updates to horizontal", () => {
        component.act(() => window.resizeTo(1200, 800));

        expect(component.tree).toMatchSnapshot();
      });
    });
  });

  it("renders custom questions", () => {
    props.practiceExplanations = [
      {button: "First Button", heading: "First Heading", text: "First Text"},
      {
        button: "Second Button",
        heading: "Second Heading",
        text: "Second Text",
        video: "https://cdn.traitify.com/images/cognitive/practice-example-h.mp4"
      }
    ];
    props.practiceQuestions = [1, 3].map((answer, index) => ({
      correctAnswerID: `r-${index}-${answer}`,
      id: `s-${index}`,
      questionImage: {url: `https://cdn.traitify.com/images/cognitive/practice-question-${index + 1}/question.png`},
      responses: [1, 2, 3, 4].map((response) => ({
        id: `r-${index}-${response}`,
        image: {url: `https://cdn.traitify.com/images/cognitive/practice-question-${index + 1}/response-${response}.png`}
      }))
    }));

    const component = new ComponentHandler(<Component {...props} />);
    expect(component.tree).toMatchSnapshot();
    expect(props.onFinish).not.toHaveBeenCalled();

    component.act(() => selectAnswer({component, index: 3}));
    expect(component.tree).toMatchSnapshot();

    component.act(() => component.findByText("First Button").props.onClick());
    expect(component.tree).toMatchSnapshot();
    expect(props.onFinish).not.toHaveBeenCalled();

    component.act(() => selectAnswer({component, index: 3}));
    expect(component.tree).toMatchSnapshot();

    component.act(() => component.findByText("Second Button").props.onClick());
    expect(component.tree).toMatchSnapshot();
    expect(props.onFinish).toHaveBeenCalled();
  });

  it("renders question 1", () => {
    const component = new ComponentHandler(<Component {...props} />);

    expect(component.tree).toMatchSnapshot();
    expect(props.onFinish).not.toHaveBeenCalled();
  });

  it("renders question 1 explanation", () => {
    const component = new ComponentHandler(<Component {...props} />);
    component.act(() => selectAnswer({component, index: 3}));

    expect(component.tree).toMatchSnapshot();
    expect(props.onFinish).not.toHaveBeenCalled();
  });

  it("renders question 2", () => {
    const component = new ComponentHandler(<Component {...props} />);
    component.act(() => selectAnswer({component, index: 2}));
    component.act(() => component.findByText("cognitive_practice_step_1_button").props.onClick());

    expect(component.tree).toMatchSnapshot();
    expect(props.onFinish).not.toHaveBeenCalled();
  });

  it("renders question 2 explanation", () => {
    const component = new ComponentHandler(<Component {...props} />);
    component.act(() => selectAnswer({component, index: 2}));
    component.act(() => component.findByText("cognitive_practice_step_1_button").props.onClick());
    component.act(() => selectAnswer({component, index: 2}));

    expect(component.tree).toMatchSnapshot();
    expect(props.onFinish).not.toHaveBeenCalled();
  });

  it("renders question 3", () => {
    const component = new ComponentHandler(<Component {...props} />);
    component.act(() => selectAnswer({component, index: 3}));
    component.act(() => component.findByText("cognitive_practice_step_1_button").props.onClick());
    component.act(() => selectAnswer({component, index: 3}));
    component.act(() => component.findByText("cognitive_practice_step_2_button").props.onClick());

    expect(component.tree).toMatchSnapshot();
    expect(props.onFinish).not.toHaveBeenCalled();
  });

  it("renders question 3 explanation", () => {
    const component = new ComponentHandler(<Component {...props} />);
    component.act(() => selectAnswer({component, index: 3}));
    component.act(() => component.findByText("cognitive_practice_step_1_button").props.onClick());
    component.act(() => selectAnswer({component, index: 3}));
    component.act(() => component.findByText("cognitive_practice_step_2_button").props.onClick());
    component.act(() => selectAnswer({component, index: 3}));

    expect(component.tree).toMatchSnapshot();
    expect(props.onFinish).not.toHaveBeenCalled();
  });
});
