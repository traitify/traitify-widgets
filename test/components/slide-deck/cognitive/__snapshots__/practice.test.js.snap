// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Practice actions finishes 1`] = `
<div
  className="loadingContainer"
>
  <div
    className="loading"
  >
    <i />
    <i />
  </div>
</div>
`;

exports[`Practice renders custom questions 1`] = `
<div
  className="practiceContainer"
>
  <div
    className="statusContainer"
  >
    <div
      className="timer"
    >
      cognitive_practice_heading
    </div>
    <div
      className="status"
    >
      1
       / 
      2
    </div>
    <div
      className="progressBar"
    >
      <div
        className="progress"
        style={
          {
            "width": "50%",
          }
        }
      />
    </div>
  </div>
  <div
    className="mock"
  >
    Slide
  </div>
</div>
`;

exports[`Practice renders custom questions 2`] = `
<div
  className="instructions"
>
  <h1>
    First Heading
  </h1>
  <p
    className="center"
  >
    cognitive_practice_answer_incorrect
  </p>
  <div
    className="text"
    dangerouslySetInnerHTML={
      {
        "__html": "First Text",
      }
    }
  />
  <button
    className="btnBlue"
    onClick={[Function]}
    type="button"
  >
    First Button
  </button>
</div>
`;

exports[`Practice renders custom questions 3`] = `
<div
  className="practiceContainer"
>
  <div
    className="statusContainer"
  >
    <div
      className="timer"
    >
      cognitive_practice_heading
    </div>
    <div
      className="status"
    >
      2
       / 
      2
    </div>
    <div
      className="progressBar"
    >
      <div
        className="progress"
        style={
          {
            "width": "100%",
          }
        }
      />
    </div>
  </div>
  <div
    className="mock"
  >
    Slide
  </div>
</div>
`;

exports[`Practice renders custom questions 4`] = `
<div
  className="instructions"
>
  <h1>
    Second Heading
  </h1>
  <p
    className="center"
  >
    cognitive_practice_answer_incorrect
  </p>
  <div
    className="text"
    dangerouslySetInnerHTML={
      {
        "__html": "Second Text",
      }
    }
  />
  <video
    autoPlay={true}
    disableremoteplayback="true"
    loop={true}
    muted={true}
    playsInline={true}
  >
    <source
      src="https://cdn.traitify.com/images/cognitive/practice-example-h.mp4"
      type="video/mp4"
    />
  </video>
  <button
    className="btnBlue"
    onClick={[Function]}
    type="button"
  >
    Second Button
  </button>
</div>
`;

exports[`Practice renders custom questions 5`] = `
<div
  className="loadingContainer"
>
  <div
    className="loading"
  >
    <i />
    <i />
  </div>
</div>
`;

exports[`Practice renders question 1 1`] = `
<div
  className="practiceContainer"
>
  <div
    className="statusContainer"
  >
    <div
      className="timer"
    >
      cognitive_practice_heading
    </div>
    <div
      className="status"
    >
      1
       / 
      3
    </div>
    <div
      className="progressBar"
    >
      <div
        className="progress"
        style={
          {
            "width": "33.333333333333336%",
          }
        }
      />
    </div>
  </div>
  <div
    className="mock"
  >
    Slide
  </div>
</div>
`;

exports[`Practice renders question 1 explanation 1`] = `
<div
  className="instructions"
>
  <h1>
    cognitive_practice_step_1_heading
  </h1>
  <p
    className="center"
  >
    cognitive_practice_answer_correct
  </p>
  <div
    className="text"
    dangerouslySetInnerHTML={
      {
        "__html": "cognitive_practice_step_1_text",
      }
    }
  />
  <video
    autoPlay={true}
    disableremoteplayback="true"
    loop={true}
    muted={true}
    playsInline={true}
  >
    <source
      src="https://cdn.traitify.com/images/cognitive/practice-1-h.mp4"
      type="video/mp4"
    />
  </video>
  <button
    className="btnBlue"
    onClick={[Function]}
    type="button"
  >
    cognitive_practice_step_1_button
  </button>
</div>
`;

exports[`Practice renders question 2 1`] = `
<div
  className="practiceContainer"
>
  <div
    className="statusContainer"
  >
    <div
      className="timer"
    >
      cognitive_practice_heading
    </div>
    <div
      className="status"
    >
      2
       / 
      3
    </div>
    <div
      className="progressBar"
    >
      <div
        className="progress"
        style={
          {
            "width": "66.66666666666667%",
          }
        }
      />
    </div>
  </div>
  <div
    className="mock"
  >
    Slide
  </div>
</div>
`;

exports[`Practice renders question 2 explanation 1`] = `
<div
  className="instructions"
>
  <h1>
    cognitive_practice_step_2_heading
  </h1>
  <p
    className="center"
  >
    cognitive_practice_answer_incorrect
  </p>
  <div
    className="text"
    dangerouslySetInnerHTML={
      {
        "__html": "cognitive_practice_step_2_text",
      }
    }
  />
  <video
    autoPlay={true}
    disableremoteplayback="true"
    loop={true}
    muted={true}
    playsInline={true}
  >
    <source
      src="https://cdn.traitify.com/images/cognitive/practice-2-h.mp4"
      type="video/mp4"
    />
  </video>
  <button
    className="btnBlue"
    onClick={[Function]}
    type="button"
  >
    cognitive_practice_step_2_button
  </button>
</div>
`;

exports[`Practice renders question 3 1`] = `
<div
  className="practiceContainer"
>
  <div
    className="statusContainer"
  >
    <div
      className="timer"
    >
      cognitive_practice_heading
    </div>
    <div
      className="status"
    >
      3
       / 
      3
    </div>
    <div
      className="progressBar"
    >
      <div
        className="progress"
        style={
          {
            "width": "100%",
          }
        }
      />
    </div>
  </div>
  <div
    className="mock"
  >
    Slide
  </div>
</div>
`;

exports[`Practice renders question 3 explanation 1`] = `
<div
  className="instructions"
>
  <h1>
    cognitive_practice_step_3_heading
  </h1>
  <p
    className="center"
  >
    cognitive_practice_answer_incorrect
  </p>
  <div
    className="text"
    dangerouslySetInnerHTML={
      {
        "__html": "cognitive_practice_step_3_text",
      }
    }
  />
  <video
    autoPlay={true}
    disableremoteplayback="true"
    loop={true}
    muted={true}
    playsInline={true}
  >
    <source
      src="https://cdn.traitify.com/images/cognitive/practice-3-h.mp4"
      type="video/mp4"
    />
  </video>
  <button
    className="btnBlue"
    onClick={[Function]}
    type="button"
  >
    cognitive_practice_step_3_button
  </button>
</div>
`;

exports[`Practice videos horizontal renders component 1`] = `
<div
  className="instructions"
>
  <h1>
    cognitive_practice_step_1_heading
  </h1>
  <p
    className="center"
  >
    cognitive_practice_answer_incorrect
  </p>
  <div
    className="text"
    dangerouslySetInnerHTML={
      {
        "__html": "cognitive_practice_step_1_text",
      }
    }
  />
  <video
    autoPlay={true}
    disableremoteplayback="true"
    loop={true}
    muted={true}
    playsInline={true}
  >
    <source
      src="https://cdn.traitify.com/images/cognitive/practice-1-h.mp4"
      type="video/mp4"
    />
  </video>
  <button
    className="btnBlue"
    onClick={[Function]}
    type="button"
  >
    cognitive_practice_step_1_button
  </button>
</div>
`;

exports[`Practice videos horizontal updates to vertical 1`] = `
<div
  className="instructions"
>
  <h1>
    cognitive_practice_step_1_heading
  </h1>
  <p
    className="center"
  >
    cognitive_practice_answer_incorrect
  </p>
  <div
    className="text"
    dangerouslySetInnerHTML={
      {
        "__html": "cognitive_practice_step_1_text",
      }
    }
  />
  <video
    autoPlay={true}
    disableremoteplayback="true"
    loop={true}
    muted={true}
    playsInline={true}
  >
    <source
      src="https://cdn.traitify.com/images/cognitive/practice-1-v.mp4"
      type="video/mp4"
    />
  </video>
  <button
    className="btnBlue"
    onClick={[Function]}
    type="button"
  >
    cognitive_practice_step_1_button
  </button>
</div>
`;

exports[`Practice videos vertical renders component 1`] = `
<div
  className="instructions"
>
  <h1>
    cognitive_practice_step_1_heading
  </h1>
  <p
    className="center"
  >
    cognitive_practice_answer_incorrect
  </p>
  <div
    className="text"
    dangerouslySetInnerHTML={
      {
        "__html": "cognitive_practice_step_1_text",
      }
    }
  />
  <video
    autoPlay={true}
    disableremoteplayback="true"
    loop={true}
    muted={true}
    playsInline={true}
  >
    <source
      src="https://cdn.traitify.com/images/cognitive/practice-1-v.mp4"
      type="video/mp4"
    />
  </video>
  <button
    className="btnBlue"
    onClick={[Function]}
    type="button"
  >
    cognitive_practice_step_1_button
  </button>
</div>
`;

exports[`Practice videos vertical updates to horizontal 1`] = `
<div
  className="instructions"
>
  <h1>
    cognitive_practice_step_1_heading
  </h1>
  <p
    className="center"
  >
    cognitive_practice_answer_incorrect
  </p>
  <div
    className="text"
    dangerouslySetInnerHTML={
      {
        "__html": "cognitive_practice_step_1_text",
      }
    }
  />
  <video
    autoPlay={true}
    disableremoteplayback="true"
    loop={true}
    muted={true}
    playsInline={true}
  >
    <source
      src="https://cdn.traitify.com/images/cognitive/practice-1-h.mp4"
      type="video/mp4"
    />
  </video>
  <button
    className="btnBlue"
    onClick={[Function]}
    type="button"
  >
    cognitive_practice_step_1_button
  </button>
</div>
`;
